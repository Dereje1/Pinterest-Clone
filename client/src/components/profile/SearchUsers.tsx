import React, { useEffect, useMemo, useState } from 'react';
import _ from 'lodash';
import { useHistory } from 'react-router-dom';
import Paper from '@mui/material/Paper';
import InputBase from '@mui/material/InputBase';
import HighlightOffIcon from '@mui/icons-material/HighlightOff';
import IconButton from '@mui/material/IconButton';
import Autocomplete from '@mui/material/Autocomplete';
import ListItem from '@mui/material/ListItem';
import ListItemAvatar from '@mui/material/ListItemAvatar';
import ListItemText from '@mui/material/ListItemText';
import Avatar from '@mui/material/Avatar';
import { getProviderIcons, ProfileLink } from '../common/common';
import { providerIconsType } from '../../interfaces';
import RESTcall from '../../crud';

const providerIcons: providerIconsType = getProviderIcons({ fontSize: 20 });

interface SearchUsersProps {
  closeSearch: () => void
  authenticated: boolean
  displayLogin: () => void
}

interface FoundUser {
  displayName: string,
  service: string,
  _id: string
}

type FoundUsers = FoundUser[] | []

function SearchUser({ closeSearch, displayLogin, authenticated }: SearchUsersProps) {
  const [searchVal, setSearchVal] = useState('');
  const [foundUsers, setFoundUsers] = useState([] as FoundUsers);
  const [isLoading, setIsLoading] = useState(false);

  const history = useHistory();

  const handleSearch = async () => {
    if (searchVal.trim().length) {
      const data: FoundUsers = await RESTcall({ address: `/api/searchUser/${searchVal}` });
      setFoundUsers(data);
      setIsLoading(false);
    } else {
      setFoundUsers([]);
      setIsLoading(false);
    }
  };

  useEffect(() => {
    if (isLoading) {
      handleSearch();
    }
  }, [isLoading]);

  const onDebounceSearch = useMemo(() => _.debounce(() => setIsLoading(true), 1000), []);

  const handleSelection = (e: React.SyntheticEvent, value: FoundUser) => {
    closeSearch();
    history.push(`/profile/${value._id}`);
  };

  return (
    <Paper
      component="form"
      sx={{
        p: '2px 4px',
        display: 'flex',
        alignItems: 'center',
        width: 300,
        marginTop: 'auto',
        marginBottom: 'auto',
      }}
      variant="elevation"
      elevation={2}
    >
      <Autocomplete
        id="free-solo-user-search"
        freeSolo
        options={foundUsers}
        getOptionLabel={(option: FoundUser) => option.displayName}
        fullWidth
        disableClearable
        inputValue={searchVal}
        onInputChange={(e, value) => setSearchVal(value)}
        onChange={handleSelection}
        onKeyUp={onDebounceSearch}
        loading={isLoading}
        renderOption={
          (props, option) => (
            <ListItem
              {...props}
              key={option._id}
            >
              <ListItemAvatar>
                <Avatar
                  sx={{
                    width: 26,
                    height: 26,
                    bgcolor: providerIcons[option.service as keyof providerIconsType].color,
                  }}
                >
                  {providerIcons[option.service as keyof providerIconsType].icon}
                </Avatar>
              </ListItemAvatar>
              <ProfileLink
                authenticated={authenticated}
                closePin={closeSearch}
                userId={option._id}
                title={(
                  <ListItemText
                    primary={option.displayName}
                    primaryTypographyProps={{ color: '#3752ff', fontWeight: 'bold' }}
                  />
                )}
                displayLogin={displayLogin}
              />
            </ListItem>
          )
        }
        renderInput={(params) => (
          <InputBase
            {...params.InputProps}
            sx={{ width: '100%', height: '100%', marginLeft: 1 }}
            placeholder="Search users..."
            inputProps={{ ...params.inputProps }}
          />
        )}
      />

      <IconButton
        type="button"
        id="clear-search"
        sx={{
          p: '10px',
          cursor: 'pointer',
        }}
        onClick={closeSearch}
      >
        <HighlightOffIcon />
      </IconButton>
    </Paper>
  );
}

export default SearchUser;
