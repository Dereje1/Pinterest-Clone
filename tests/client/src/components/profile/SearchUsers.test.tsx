import React from 'react';
import * as router from 'react-router-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import SearchUsers from '../../../../../client/src/components/profile/SearchUsers';
import RESTcall from '../../../../../client/src/crud';

jest.mock('../../../../../client/src/crud');
const mockedRESTcall = jest.mocked(RESTcall);
jest.useFakeTimers();
// Mock router hooks
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
  useHistory: jest.fn(),
}));

describe('The Search users component', () => {
  let props: React.ComponentProps<typeof SearchUsers>;
  beforeEach(() => {
    props = {
      authenticated: true,
      closeSearch: jest.fn(),
    };
  });

  afterEach(() => {
    mockedRESTcall.mockClear();
  });

  test('will render the component', () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will render the input', () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    const autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    const input = autoComplete.props().renderInput({ InputProps: '' });
    const { sx, placeholder } = input.props;
    expect({ sx, placeholder }).toEqual({
      sx: { width: '100%', height: '100%', marginLeft: 1 },
      placeholder: 'Search users...',
    });
  });

  test('will prevent default on enter in input', () => {
    const preventDefault = jest.fn();
    const wrapper = shallow(<SearchUsers {...props} />);
    const autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    const input = autoComplete.props().renderInput({ InputProps: '' });
    input.props.inputProps.onKeyDown({ key: 'Enter', preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });

  test('will render a single option', () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    const autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    let option = autoComplete.props().renderOption(
      { },
      { _id: 'test user Id', service: 'twitter', displayName: 'test displayName' },
    );
    expect(option.props.service).toBe('twitter');
    expect(option.props.profileLinkProps.userId).toBe('test user Id');
    expect(option.props.profileLinkProps.title.props.primary).toBe('test displayName');
    option = autoComplete.props().renderOption(
      { },
      'a string',
    );
    expect(option).toBe(null);
  });

  test('will get the option label', () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    const autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    let optionLabel = autoComplete.props().getOptionLabel({ _id: 'test user Id', service: 'twitter', displayName: 'test displayName' });
    expect(optionLabel).toBe('test displayName');
    optionLabel = autoComplete.props().getOptionLabel('a string');
    expect(optionLabel).toBe('');
  });

  test('will update the controlled input', () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    let autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    expect(autoComplete.props().inputValue).toBe('');
    autoComplete.props().onInputChange('', 'new value');
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    expect(autoComplete.props().inputValue).toBe('new value');
  });

  test('will make a rest call on keyUp and set the options', async () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    let autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // first change input
    autoComplete.props().onInputChange('', 'search-user');
    // update after input change
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // call keyup
    autoComplete.props().onKeyUp({ key: '' });
    // update again after keyup
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // advance for debounce and resolve crud call
    jest.advanceTimersByTime(1000);
    await Promise.resolve();
    // update again after rest call resolves
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    expect(mockedRESTcall).toHaveBeenCalledWith({ address: '/api/searchUser/search-user' });
    expect(autoComplete.props().options).toEqual([
      {
        displayName: 'test_displayName',
        service: 'twitter',
        _id: 'test_user_id',
      },
    ]);
    expect(autoComplete.props().loading).toBe(false);
  });

  test('will not make a rest call on keyUp if search field is empty', async () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    let autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // first change input
    autoComplete.props().onInputChange('', '   ');
    // update after input change
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // call keyup
    autoComplete.props().onKeyUp({ key: '' });
    // update again after keyup
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // advance for debounce
    jest.advanceTimersByTime(1000);
    // update again after rest call resolves
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    expect(mockedRESTcall).toHaveBeenCalledTimes(0);
    expect(autoComplete.props().options).toEqual([]);
    expect(autoComplete.props().loading).toBe(false);
  });

  test('will not make a rest call on ArrowDown of key up', async () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    let autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // first change input
    autoComplete.props().onInputChange('', 'search-user');
    // update after input change
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // call keyup
    autoComplete.props().onKeyUp({ key: 'ArrowDown' });
    // update again after keyup
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // advance for debounce
    jest.advanceTimersByTime(1000);
    // update again after rest call resolves
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    expect(mockedRESTcall).toHaveBeenCalledTimes(0);
    expect(autoComplete.props().options).toEqual([]);
    expect(autoComplete.props().loading).toBe(false);
  });

  test('will not make a rest call on ArrowUp of key up', async () => {
    const wrapper = shallow(<SearchUsers {...props} />);
    let autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // first change input
    autoComplete.props().onInputChange('', 'search-user');
    // update after input change
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // call keyup
    autoComplete.props().onKeyUp({ key: 'ArrowUp' });
    // update again after keyup
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    // advance for debounce
    jest.advanceTimersByTime(1000);
    // update again after rest call resolves
    autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    expect(mockedRESTcall).toHaveBeenCalledTimes(0);
    expect(autoComplete.props().options).toEqual([]);
    expect(autoComplete.props().loading).toBe(false);
  });

  test('will update router history on selection of a user', () => {
    const push = jest.fn();
    const hist = router.useHistory();
    jest
      .spyOn(router, 'useHistory')
      .mockImplementation(() => ({ ...hist, push }));

    const wrapper = shallow(<SearchUsers {...props} />);
    const autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    autoComplete.props().onChange({}, { _id: 'a_user_id' });
    expect(props.closeSearch).toHaveBeenCalledTimes(1);
    expect(push).toHaveBeenCalledWith('/profile/a_user_id');
  });

  test('will not reroute on create option', () => {
    const push = jest.fn();
    const hist = router.useHistory();
    jest
      .spyOn(router, 'useHistory')
      .mockImplementation(() => ({ ...hist, push }));

    const wrapper = shallow(<SearchUsers {...props} />);
    const autoComplete = wrapper.find({ id: 'free-solo-user-search' });
    autoComplete.props().onChange({}, { _id: 'a_user_id' }, 'createOption');
    expect(props.closeSearch).toHaveBeenCalledTimes(0);
    expect(push).toHaveBeenCalledTimes(0);
  });
});
