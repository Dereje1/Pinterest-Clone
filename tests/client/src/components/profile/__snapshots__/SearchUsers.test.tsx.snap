// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Search users component will render the component 1`] = `
<ForwardRef(Paper)
  component="form"
  elevation={2}
  sx={
    {
      "alignItems": "center",
      "display": "flex",
      "marginBottom": "auto",
      "marginTop": "auto",
      "p": "2px 4px",
      "width": 300,
    }
  }
  variant="elevation"
>
  <ForwardRef(Autocomplete)
    disableClearable={true}
    freeSolo={true}
    fullWidth={true}
    getOptionLabel={[Function]}
    id="free-solo-user-search"
    inputValue=""
    loading={false}
    onChange={[Function]}
    onInputChange={[Function]}
    onKeyUp={[Function]}
    options={[]}
    renderInput={[Function]}
    renderOption={[Function]}
  />
  <ForwardRef(IconButton)
    id="clear-search"
    onClick={[MockFunction]}
    sx={
      {
        "cursor": "pointer",
        "p": "10px",
      }
    }
    type="button"
  >
    <Memo(ForwardRef(HighlightOffIcon)) />
  </ForwardRef(IconButton)>
</ForwardRef(Paper)>
`;
